README.md

```diff
@@ -131,6 +131,22 @@ enum.next #=> ["It"]
 enum.next #=> ["is"]
 ```
 
+On versions later than 1.4.0, you can use `RE2::Set` to match multiple patterns
+against a string. Calling `RE2::Set#add` with a pattern will return an integer
+index of the pattern. After all patterns have been added, the set can be
+compiled using `RE2::Set#compile`, and then `RE2::Set#match` will return an
+`Array<Integer>` containing the indexes of all the patterns that matched.
+
+``` ruby
+set = RE2::Set
+set.add("abc") #=> 0
+set.add("def") #=> 1
+set.add("ghi") #=> 2
+set.compile #=> true
+set.match("abcdefghi") #=> [0, 1, 2]
```

Comment:
This closely matches RE2's own API (`Add`, `Compile` and `Match`) but I wonder if the fact a user has to call `compile` before calling `match` (and that calling `add` after `compile` will raise an error) is surprising in Ruby.

I suppose it's similar to freezing strings (which raises an error if you attempt mutate a string after calling `freeze`) but would it be worth having separate classes for uncompiled Sets and compiled Sets, e.g.

```ruby
set = RE2::Set.new
#=> #<RE2::Set ...>
set.add("foo")
set.match("foooo")
#=> NoMethodError: undefined method `match' for #<RE2::Set ...>
set.compile
#=> #<RE2::CompiledSet ...>
set.add("foo")
#=> NoMethodError: undefined method `add' for #<RE2::CompiledSet ...>
set.match("fooooo")
#=> [0, 1, 2]
```

This way we're enforcing the correct usage at the API level rather than with runtime errors. That said, perhaps the runtime errors are more informative?

Reply:
I don't think closely matching RE2's own API is too surprising in Ruby, but then again I don't have a lot of experience writing Ruby code, only reading.

However, having separate classes like you suggest seems, to me, more surprising. I'm assuming you meant that `Set#compile` returns a new `CompiledSet` but does not change `self` from an object of `Set` into an object of `CompiledSet`, since Ruby provides no way (that I could find) to change the class of an object. This raises two issues:

1. `RE2::Set` is now no longer an `RE2::Set`, since you can't actually use it to match regexps. At best, it is sth more like an `RE2::SetBuilder`.
2. Even if we rename `RE2::Set` to `RE2::SetBuilder` (or sth like that), its utility would be immediately lost when `#compile` is called on it. Or, we make it capable of adding new patterns after `#compile`, so a new `CompiledSet` may be returned from a subsequent call to `#compile` that includes both the initial and subsequent patterns.
3. If we were to add the ability to generate multiple `CompiledSet`s from a single `SetBuilder`, we'd need to save every pattern that's been added to a `SetBuilder` instance, and within every call to `#compile`, create a new C++ `RE2::Set` object internally and re-add all previous patterns to it. Any call to `#compile` at any time could then compile this internal C++ `RE2::Set` and return it.

But, that doesn't seem very idiomatic Ruby anyway, and there exists a better option:

```ruby
module RE2
  # @example
  #    set = RE2::Set(["abc", "def", "ghi"])
  #    set.match("abcdefghi") => [0, 1, 2]
  def Set(patterns, anchor = :unanchored, options = {})
    set = RE2::Set.new(anchor, options)
    patterns.each { |p| set.add(p) }
    set.compile
    set
  end
end
```

----

My best reason to stick to C++ RE2's own API is that it results in a clean mapping for users coming from C++ (like me). My second best reason is that I already have code written in this style. Of course, I have no issues against making a top-level function that takes a list of patterns to add and compile in a single-shot. But I'd prefer to not hide the internals for anyone else like me to also be able to use it, if they so please.

Reply:
That makes sense and your point about a `RE2::SetBuilder` becoming useless after calling `compile` (without reworking its internals to manage multiple underlying `RE2::Set`s) is a good one. I was hoping to avoid the mutable state of a single instance by having a more functional interface but this shows we can't really escape the need for one object to become "compiled" and therefore unusable.

Keeping things similar to the underlying C++ API seems like the best course of action.